crates/ide-assists/src/handlers/bool_to_enum.rs

```diff
@@ -0,0 +1,1514 @@
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: bool_to_enum
+//
+// This converts boolean local variables, fields, constants, and statics into a new
+// enum with two variants `Bool::True` and `Bool::False`, as well as replacing
+// all assignments with the variants and replacing all usages with `== Bool::True` or
+// `== Bool::False`.
+//
+// ```
+// fn main() {
+//     let $0bool = true;
+//
+//     if bool {
+//         println!("foo");
+//     }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     #[derive(PartialEq, Eq)]
+//     enum Bool { True, False }
+//
+//     let bool = Bool::True;
+//
+//     if bool == Bool::True {
+//         println!("foo");
+//     }
+// }
+// ```
+pub(crate) fn bool_to_enum(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let BoolNodeData { target_node, name, ty_annotation, initializer, definition } =
+        find_bool_node(ctx)?;
+    let target_module = ctx.sema.scope(&target_node)?.module().nearest_non_block_module(ctx.db());
+
+    let target = name.syntax().text_range();
+    acc.add(
+        AssistId("bool_to_enum", AssistKind::RefactorRewrite),
+        "Convert boolean to enum",
+        target,
+        |edit| {
+            if let Some(ty) = &ty_annotation {
+                cov_mark::hit!(replaces_ty_annotation);
+                edit.replace(ty.syntax().text_range(), "Bool");
+            }
+
+            if let Some(initializer) = initializer {
+                replace_bool_expr(edit, initializer);
+            }
+
+            let usages = definition.usages(&ctx.sema).all();
+            add_enum_def(edit, ctx, &usages, target_node, &target_module);
+            replace_usages(edit, ctx, &usages, definition, &target_module);
+        },
+    )
+}
+
+struct BoolNodeData {
+    target_node: SyntaxNode,
+    name: ast::Name,
+    ty_annotation: Option<ast::Type>,
+    initializer: Option<ast::Expr>,
+    definition: Definition,
+}
+
+/// Attempts to find an appropriate node to apply the action to.
+fn find_bool_node(ctx: &AssistContext<'_>) -> Option<BoolNodeData> {
+    let name: ast::Name = ctx.find_node_at_offset()?;
+
+    if let Some(let_stmt) = name.syntax().ancestors().find_map(ast::LetStmt::cast) {
+        let bind_pat = match let_stmt.pat()? {
+            ast::Pat::IdentPat(pat) => pat,
+            _ => {
+                cov_mark::hit!(not_applicable_in_non_ident_pat);
+                return None;
+            }
+        };
+        let def = ctx.sema.to_def(&bind_pat)?;
+        if !def.ty(ctx.db()).is_bool() {
+            cov_mark::hit!(not_applicable_non_bool_local);
+            return None;
+        }
+
+        Some(BoolNodeData {
+            target_node: let_stmt.syntax().clone(),
+            name,
+            ty_annotation: let_stmt.ty(),
+            initializer: let_stmt.initializer(),
+            definition: Definition::Local(def),
+        })
+    } else if let Some(const_) = name.syntax().ancestors().find_map(ast::Const::cast) {
```

Comment:
```suggestion
    } else if let Some(const_) = name.syntax().parent().and_then(ast::Const::cast) {
```

Here we only need the parent, as the name is the direct child of the item.